# Combinatorics: Counting Events
- Use to calculate how many possible options there are
- Useful for event probability problems to count numerator or denominator

**Questions to ask:**
- Are we arranging or selecting items?
- Ordered or unordered?
- With or without replacement?

### Summary Table
- Note: Factorial: $n! = n \cdot (n-1) \cdot (n-2) \cdots 1, \quad 0! = 1$

| Case                          | Ordered? | Replacement? | Formula |
|-------------------------------|----------|--------------|---------|
| Combinations (no repeats)     | No       | No           | $C(n,r) = \binom{n}{r} = \frac{n!}{r!(n-r)!}$ |
| Combinations (with repeats)   | No       | Yes          | $C(n+r-1,r) = \binom{n+r-1}{r}$ |
| Permutations (no repeats)     | Yes      | No           | $P(n,r) = \frac{n!}{(n-r)!}$ |
| Permutations (with repeats)   | Yes      | Yes          | $n^r$ |

### Determine Which Formula:
Is order important?
- No → **Combinations** 
  - **Without replacement:**
    - Example: Card hand (5 cards from 52)
    - $n = 52, r = 5$
    - $\binom{52}{5} = \frac{52!}{5!(52-5)!} = \tfrac{52!}{5!47!}$
  - **With replacement:**
    - Example: Ice cream scoops (3 scoops, 5 flavors, repeats allowed)
    - $n = 5, r = 3$
    - $\tbinom{5+3-1}{3} = \tbinom{7}{3} = 35$ 
- Yes → **Permutations**
  - **Without replacement:**  
    - Example: Race rankings (3 winners out of 10 runners, no repeats)
    - $n = 10, r = 3$
    - $\tfrac{10!}{10-3!} = \tfrac{10!}{7!} = 720$
  - **With replacement:**  
    - Example: PIN code (4 digits, repeats allowed)
    - $10^4 = 10,000$  

Takeaways:
- Binomial: $\binom{n}{r}$
  - Read as **“n choose r”**.  
  - Represents the number of ways to choose $r$ items from $n$ total items **without regard to order**.  
- Factorial: $n!$
  - Factorials appear whenever “arranging” or “removing order” is involved.
- Exponents: $n^r$
  - Plain powers only apply when order matters and repeats are allowed.

In [11]:
def factorial(n):
    """Calculate the factorial of n (n!) for non-negative integers."""
    if n < 0:
        raise ValueError("Factorial is not defined for negative numbers.")
    result = 1
    for i in range(2, n + 1):
        result *= i
    return result

print(factorial(5))  # Example usage: 5! = 120

120


In [7]:
def n_choose_r(n, r):
    """Calculate the binomial coefficient (n choose r)."""
    if not (0 <= r <= n):
        return 0
    return factorial(n) // (factorial(r) * factorial(n - r))

print(n_choose_r(52, 5))  # Example usage: 5 choose 2 = 10

2598960


In [16]:
def permutations(n, r):
    """Calculate the number of permutations of n items taken r at a time."""
    if not (0 <= r <= n):
        return 0
    return factorial(n) // factorial(n - r)

### 1. Combinations without Replacement
- Unordered selection, no repeats
- Number of ways to choose $r$ items from $n$, order ignored:  
$$
C(n,r) = \binom{n}{r} = \frac{n!}{r!(n-r)!}
$$

**Problem:**  
From a standard deck of 52 cards, 5 cards are drawn at random.  
What is the probability that all 5 cards are **hearts**?

**Solution:**

- Total number of ways to choose 5 cards from 52: $\binom{52}{5}$
- Number of favorable ways (all hearts): $\binom{13}{5}$

$$
P(\text{5 hearts}) = \frac{\binom{13}{5}}{\binom{52}{5}}
$$

$$
P(\text{5 hearts}) = \frac{1287}{2,598,960} \approx 0.000495
$$

In [ ]:
combo_5_hearts = n_choose_r(13, 5) # Number of combinations of 5 hearts from 13 hearts
combo_5_cards = n_choose_r(52, 5) # Number of cominations of 5 cards from 52 cards
prob_5_hearts = combo_5_hearts / combo_5_cards
print(prob_5_hearts)  # Probability of drawing 5 hearts from a deck of 52 cards

0.0004951980792316927


### 2. Combinations with Repacement
- Unordered selection with repeats
  - a.k.a. “stars and bars”
- Number of ways to choose $r$ items from $n$ categories, order ignored: 
$$
C(n+r-1,r) = \binom{n+r-1}{r}
$$

**Example: Combinations with Replacement**

**Problem:**  
An ice cream shop offers 5 flavors. A customer orders a cup with 3 scoops of ice cream.  
Scoops can be of the same flavor or different flavors, and the order of scoops does not matter.  
What is the probability that all 3 scoops are of the **same flavor**?

**Solution:**

- $n = 5, r = 3$
- Total number of unordered selections with replacement:  

$$
\binom{n+r-1}{r} = \binom{5+3-1}{3} = \binom{7}{3}
$$

$$
\binom{7}{3} = \frac{7!}{3!(4!)} = 35
$$

$$
\text{There are 35 total possible combinations}
$$


- Favorable outcomes (all scoops the same):  
    - There are **5 ways** (one for each flavor)

- Probability:  

$$
P(\text{all 3 same}) = \frac{5}{35} = \frac{1}{7} \approx 0.143
$$

In [13]:
combo_scoops = n_choose_r(7, 3)
combo_all_same = 5
prob_all_same = combo_all_same / combo_scoops
print(prob_all_same)  # Probability of all 3 scoops being the same flavor

0.14285714285714285


### 3. Permutations - No Replacement

- Ordered Selection, no repeats
- Number of ways to arrange $r$ items from $n$  
  $$
  P(n,r) = \frac{n!}{(n-r)!}
  $$

**Example: Permutation Probability with 0 Interpreted as Leading Zero**

**Problem:**  
Digits 0 through 9 are placed in a box.  
Two digits are drawn **without replacement** and arranged in order.  
If the first digit is 0, the number is interpreted as a 1-digit number using only the second digit (e.g., drawing 0 then 5 = the number 5).  
What is the probability that the resulting number is **greater than 76**?

**Solution:**

- **Step 1. Total possible outcomes:**
$$P(10,2) = \frac{10!}{(10-2)!} = 90$$

- **Step 2. Favorable outcomes:**
We need numbers greater than 76.

  - **Case A. First digit = 0–6**  
    - Numbers are in ranges 1–69.  
    - None exceed 76.  
    - Favorable = 0

  - **Case B. First digit = 7**  
    - Possible second digits: 0–9 except 7.  
    - Numbers range from 70–79 but not 77.  
    - Greater than 76 → {78, 79}.  
    - **Favorable = 2**

  - **Case C. First digit = 8**  
    - Possible second digits: 0–9 except 8.  
    - Numbers: 80–89 except 88.  
    - All are > 76.  
    - **Favorable = 9**

  - **Case D. First digit = 9**  
    - Possible second digits: 0–9 except 9.
    - Numbers: 90–99 except 99.
    - All are > 76.
    - **Favorable = 9**

$$
\text{favorable outcomes} = 2 + 9 + 9 = 20
$$

- **Step 3. Probability**
$$
P(\text{Number > 76}) = \frac{20}{90} = \frac{2}{9} \approx 0.222
$$

In [ ]:
combo_all = permutations(10, 2)

combo_favorable_with_first_7 = 2 # {78, 79}, no 77
combo_favorable_with_first_8 = 9 # {80, 81, 82, 83, 84, 85, 86, 87, 89}, no 88
combo_favorable_with_first_9 = 9 # {90, 91, 92, 93, 94, 95, 96, 97, 98}, no 99
combo_favorable = (
    combo_favorable_with_first_7 
    + combo_favorable_with_first_8 
    + combo_favorable_with_first_9)

prob_favorable = combo_favorable / combos_tickets

print(combo_favorable)
print(combo_all)
print(prob_favorable)

20
90
0.2222222222222222


### 3.1 Permutation Special Case: Factorial
**Number of ways to arrange $n$ distinct objects**
- Factorials can be used to solve a **special case** of combinatorics problems
- Arranging distinct objects means that **order matters** and **no replacement** → permutation
  - $P(n,r) = \tfrac{n!}{n-r!}$
- $n = r$, because arranging all items in a set
  - $\tfrac{n!}{n-r!} = \tfrac{n!}{0!} = \tfrac{n!}{1} = n!$

$$n! = n \cdot (n-1) \cdot (n-2) \cdots 1, \quad 0! = 1$$



**Factorial Question**

**Problem:**  
A password must consist of **all 6 distinct letters** from the set {A, B, C, D, E, F}, arranged in any order.  
How many unique passwords can be created?

---

**Solution:**  
Since all 6 letters are used and order matters, the number of arrangements is given by the factorial:

$$
6! = 6 \cdot 5 \cdot 4 \cdot 3 \cdot 2 \cdot 1
$$

$$
6! = 720
$$


In [ ]:
print(factorial(6))

720


### 4. Permutations with Replacement
- **Ordered Selection, with repeats:**  
- Each choice has $n$ possibilities, repeated $r$ times:  
$$
n^r
$$

**Example: Permutation Probability With Replacement**

**Problem:**  
A 3-digit PIN code is generated by choosing digits 0–9, with each digit chosen independently and digits **can repeat**.  
What is the probability that the PIN code contains **at least one 7**?

### Example: Permutation Probability With Replacement

**Problem:**  
A 3-digit PIN code is generated by choosing digits 0–9, with each digit chosen independently and digits **can repeat**.  
What is the probability that the PIN code contains **at least one 7**?

**Solution:**

- **Step 1. Total number of possible PIN codes:**
    - Since each digit has 10 possibilities and repeats are allowed:

$$
\text{Total outcomes} = 10^3 = 1000
$$

- **Step 2. Use complement rule:**

- Instead of counting directly, find the probability that no digit is a 7.
    - Each digit has 9 choices (anything but 7).  
    - For 3 digits:

$$
\text{Outcomes with no 7} = 9^3 = 729
$$

- **Step 3. Probability of at least one 7:**

$$
P(\text{Outcomes with at least one 7}) = 1 - \frac{729}{1000}
$$

$$
= 1 - \frac{729}{1000} = \frac{271}{1000} \approx{0.271}
$$

In [ ]:
# Count total combinations
num_digits = 10
pin_length = 3
combo_digits = num_digits ** pin_length

# Count combinations with no 7
non_7_digits = 9
combo_no_7 = non_7_digits ** pin_length

# Calculate probabilities
prob_no_7 = combo_no_7 / combo_digits
prob_7 = 1 - prob_no_7

print(combo_no_7)
print(combo_digits)
print(prob_no_7)
print(prob_7)

729
1000
0.729
0.271
